contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,633 @@
+	// MaxRecommended is the recommended maximum amount of resources.
+	MaxRecommended api.ResourceRequirements
+}
+```
+
+### Admission Controller ###
+
+VPA Admission Controller intercepts Pod creation requests. If the Pod is matched
+by a VPA config with mode not set to “off”, the controller rewrites the request
+by applying recommended resources to the Pod spec. Otherwise it leaves the Pod
+spec unchanged.
+
+The controller gets the recommended resources by fetching
+/recommendedPodResources from the Recommender. If the call times out or fails,
+the controller falls back to the recommendation cached in the VPA object.
+If this is also not available the controller lets the request pass-through
+with originally specified resources.
+
+Note: in future it will be possible to (optionally) enforce using VPA by marking
+the Pod as "requiring VPA". This will disallow scheduling the Pod before a
+corresponding VPA config is created. The Admission Controller will reject such
+Pods if it finds no matching VPA config. This ability will be convenient for the
+user who wants to create the VPA config together with submitting the Pod.
+
+The VPA Admission Controller will be implemented as an
+[External Admission Hook](https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks).
+Note however that this depends on the proposed feature to allow
+[mutating webhook admission controllers](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission_control_extension.md#future-work).
+
+### Recommender ###
+Recommender is the main component of the VPA. It is responsible for
+computing recommended resources. On startup the recommender fetches
+historical resource utilization of all Pods (regardless of whether
+they use VPA) together with the history of Pod OOM events from the
+History Storage. It aggregates this data and keeps it in memory.
+
+During normal operation the recommender consumes real time updates of
+resource utilization and new events via the Metrics API from
+the [Metrics Server](https://github.com/kubernetes-incubator/metrics-server).
+Additionally it watches all Pods and all VPA objects in the
+cluster. For every Pod that is matched by some VPA selector the
+Recommender computes the recommended resources and sets the
+recommendation on the VPA object.
+
+It is important to realize that one VPA object has one recommendation.
+The user is expected to use one VPA to control Pods with similar
+resource usage patterns, typically a group of replicas or shards of
+a single workload.
+
+The Recommender acts as an
+[extension-apiserver](https://kubernetes.io/docs/concepts/api-extension/apiserver-aggregation/),
+exposing a synchronous RPC interface that takes a Pod Spec and the Pod metadata
+and returns recommended resources.
+
+#### Recommender API ####
+
+```GET /recommendedPodResources```
+
+Request body:
+```go
+type GetRecommendedPodResourcesRequest {
```

Comment:
Since this is an aggregated API, it needs to (at least mostly) follow Kube API conventions.  Look at how the TokenReview and SubjectAccessReview APIs work

Reply:
Done.